rust/agama-server/src/software/license.rs

```diff
@@ -0,0 +1,264 @@
+/// "license.es.txt", "license.zh_CH.txt", etc.).
+#[derive(Clone)]
+pub struct LicensesRepo {
+    /// Repository path.
+    pub path: std::path::PathBuf,
+    /// Licenses in the repository.
+    pub licenses: Vec<License>,
+}
+
+impl LicensesRepo {
+    pub fn new<P: AsRef<Path>>(path: P) -> Self {
+        Self {
+            path: path.as_ref().to_owned(),
+            licenses: vec![],
+        }
+    }
+
+    /// Reads the licenses from the repository.
+    pub fn read(&mut self) -> Result<(), std::io::Error> {
+        let entries = read_dir(self.path.as_path())?;
+
+        for entry in entries {
+            let entry = entry?;
+            if entry.file_type()?.is_dir() {
+                let Ok(id) = entry.file_name().into_string() else {
+                    continue;
+                };
+                let license = License {
+                    id,
+                    languages: Self::find_translations(&entry.path())?,
+                };
+                self.licenses.push(license);
+            }
+        }
+        Ok(())
+    }
+
+    /// Finds a license with the given ID and language.
+    ///
+    /// If a translation is not found for the given language, it returns the default text.
+    pub fn find(&self, id: &str, language: &LanguageTag) -> Option<LicenseContent> {
+        let mut candidates: Vec<String> = vec![];
+        if let Some(territory) = &language.territory {
+            candidates.push(format!("license.{}_{}.txt", language.language, territory));
+        }
+        candidates.push(format!("license.{}.txt", language.language));
+        candidates.push("license.txt".to_string());
+
+        let license_path = candidates
+            .into_iter()
+            .map(|p| self.path.join(id).join(p))
+            .find(|p| p.exists())?;
+
+        let body: String = std::fs::read_to_string(license_path).ok()?;
+
+        Some(LicenseContent {
+            id: id.to_string(),
+            body,
+        })
+    }
+
+    /// Finds translations in the given directory.
+    ///
+    /// * `path`: directory to search translations.
+    fn find_translations(path: &PathBuf) -> Result<Vec<LanguageTag>, std::io::Error> {
+        let entries = read_dir(path).unwrap().filter_map(|entry| entry.ok());
+
+        let files = entries
+            .filter(|entry| entry.file_type().is_ok_and(|f| f.is_file()))
+            .filter_map(|entry| {
+                let path = entry.path();
+                let file = path.file_name()?;
+                file.to_owned().into_string().ok()
+            });
+
+        Ok(files
+            .filter_map(|f| Self::language_tag_from_file(&f))
+            .collect())
+    }
+
+    /// Returns the language tag for the given file.
+    ///
+    /// The language is inferred from the file name (e.g., "es-ES" for license.es_ES.txt").
+    fn language_tag_from_file(name: &str) -> Option<LanguageTag> {
+        if !name.starts_with("license") {
+            return None;
```

Comment:
this basically means there is some mess in our directory. Maybe we should also log here.